src/rust/engine/fs/src/store.rs

```diff
@@ -535,32 +544,107 @@ mod remote {
 
   use bazel_protos;
   use boxfuture::{Boxable, BoxFuture};
-  use futures::{future, Future, Stream};
+  use digest::{Digest as DigestTrait, FixedOutput};
+  use futures::{self, future, Future, Sink, Stream};
   use grpcio;
+  use sha2::Sha256;
   use std::sync::Arc;
+  use std::time::Duration;
 
   use hash::Fingerprint;
 
   #[derive(Clone)]
   pub struct ByteStore {
-    address: String,
+    cas_address: String,
     env: Arc<grpcio::Environment>,
+    chunk_size_bytes: usize,
+    upload_timeout: Duration,
   }
 
   impl ByteStore {
-    pub fn new(address: String) -> ByteStore {
-      // TODO: Pass through a parallelism configuration option here.
+    pub fn new(
+      cas_address: String,
+      thread_count: usize,
+      chunk_size_bytes: usize,
+      upload_timeout: Duration,
+    ) -> ByteStore {
+      let env = Arc::new(grpcio::Environment::new(thread_count));
       ByteStore {
-        address: address,
-        env: Arc::new(grpcio::Environment::new(1)),
+        cas_address,
+        env,
+        chunk_size_bytes,
+        upload_timeout,
       }
     }
   }
 
 
   impl super::ByteStore for ByteStore {
-    fn store_bytes(&self, _entry_type: EntryType, _bytes: Vec<u8>) -> BoxFuture<Digest, String> {
-      unimplemented!()
+    fn store_bytes(&self, _entry_type: EntryType, bytes: Vec<u8>) -> BoxFuture<Digest, String> {
+      let mut hasher = Sha256::default();
+      hasher.input(&bytes);
+      let fingerprint = Fingerprint::from_bytes_unsafe(hasher.fixed_result().as_slice());
+      let len = bytes.len();
+      let resource_name = format!("{}/uploads/{}/blobs/{}/{}", "", "", fingerprint, len);
+
+      let channel = grpcio::ChannelBuilder::new(self.env.clone()).connect(&self.cas_address);
+      // TODO: Maybe keep one reference to a client on the ByteStore, rather than newing one up for
+      // every operation.
+      let client = bazel_protos::bytestream_grpc::ByteStreamClient::new(channel);
+      match client.write_opt(grpcio::CallOption::default().timeout(self.upload_timeout)) {
+        Err(err) => {
+          future::err(format!(
+            "Error attempting to connect to upload fingerprint {}: {:?}",
+            fingerprint,
+            err
+          )).to_boxed() as BoxFuture<_, _>
+        }
+        Ok((sender, receiver)) => {
+          let mut offset = 0 as usize;
+          let stream = futures::stream::iter_ok::<_, grpcio::Error>(bytes)
```

Comment:
@stuhood I'd like to file a bug against Rust for this line; can you give me a hand working out exactly what the bug should be? 

(I ended up spending about three hours tracking this down today):

If I write `iter_ok(bytes)` instead of `iter_ok::<_, grpcio::Error>(bytes)`, the error I get is: 

```
error[E0282]: type annotations needed
   --> src/store.rs:625:29
    |   
604 |           let stream = futures::stream::iter_ok(bytes)
    |               ------ consider giving `stream` a type
...
624 |             .and_then(move |_| {
    |                             ^ consider giving this closure parameter a type

error: aborting due to previous error
```

The pointer to line 624 is, it turns out, unrelated.

I tried to add explicit types to the `stream` variable. First I tried to ascibe it: 
```
let stream: Stream<Item=(bazel_protos::bytestream::WriteRequest, grpcio::WriteFlags), Error=grpcio::Error> = futures::stream::iter_ok(bytes)
//  .chunks(...
```

which complaints to me that both:
1. expected Stream but found Map (which implements Stream)
2. Stream is not Sized, and all local variables must be Sized

so I started boxing, which gave me the very oblique:
```
error[E0599]: no method named `to_boxed` found for type `futures::AndThen<futures::AndThen<futures::Join<futures::FutureResult<bazel_protos::bytestream_grpc::ByteStreamClient, std::string::String>, futures::MapErr<futures::sink::SendAll<grpcio::StreamingCallSink<bazel_protos::bytestream::WriteRequest>, std::boxed::Box<futures::Stream<Error=grpcio::Error, Item=(bazel_protos::bytestream::WriteRequest, grpcio::WriteFlags)>>>, [closure@src/store.rs:617:51: 623:14 fingerprint:_]>>, futures::MapErr<grpcio::ClientCStreamReceiver<bazel_protos::bytestream::WriteResponse>, [closure@src/store.rs:625:32: 631:16 fingerprint:_]>, [closure@src/store.rs:624:23: 632:14 receiver:_, fingerprint:_]>, std::result::Result<store::Digest, std::string::String>, [closure@src/store.rs:633:23: 644:14 len:_, fingerprint:_]>` in the current scope
   --> src/store.rs:645:14
    |
645 |             .to_boxed()
    |              ^^^^^^^^
    |
    = note: the method `to_boxed` exists but the following trait bounds were not satisfied:
            `futures::AndThen<futures::AndThen<futures::Join<futures::FutureResult<bazel_protos::bytestream_grpc::ByteStreamClient, std::string::String>, futures::MapErr<futures::sink::SendAll<grpcio::StreamingCallSink<bazel_protos::bytestream::WriteRequest>, std::boxed::Box<futures::Stream<Error=grpcio::Error, Item=(bazel_protos::bytestream::WriteRequest, grpcio::WriteFlags)>>>, [closure@src/store.rs:617:51: 623:14 fingerprint:_]>>, futures::MapErr<grpcio::ClientCStreamReceiver<bazel_protos::bytestream::WriteResponse>, [closure@src/store.rs:625:32: 631:16 fingerprint:_]>, [closure@src/store.rs:624:23: 632:14 receiver:_, fingerprint:_]>, std::result::Result<store::Digest, std::string::String>, [closure@src/store.rs:633:23: 644:14 len:_, fingerprint:_]> : boxfuture::Boxable`
            `&futures::AndThen<futures::AndThen<futures::Join<futures::FutureResult<bazel_protos::bytestream_grpc::ByteStreamClient, std::string::String>, futures::MapErr<futures::sink::SendAll<grpcio::StreamingCallSink<bazel_protos::bytestream::WriteRequest>, std::boxed::Box<futures::Stream<Error=grpcio::Error, Item=(bazel_protos::bytestream::WriteRequest, grpcio::WriteFlags)>>>, [closure@src/store.rs:617:51: 623:14 fingerprint:_]>>, futures::MapErr<grpcio::ClientCStreamReceiver<bazel_protos::bytestream::WriteResponse>, [closure@src/store.rs:625:32: 631:16 fingerprint:_]>, [closure@src/store.rs:624:23: 632:14 receiver:_, fingerprint:_]>, std::result::Result<store::Digest, std::string::String>, [closure@src/store.rs:633:23: 644:14 len:_, fingerprint:_]> : boxfuture::Boxable`
            `&mut futures::AndThen<futures::AndThen<futures::Join<futures::FutureResult<bazel_protos::bytestream_grpc::ByteStreamClient, std::string::String>, futures::MapErr<futures::sink::SendAll<grpcio::StreamingCallSink<bazel_protos::bytestream::WriteRequest>, std::boxed::Box<futures::Stream<Error=grpcio::Error, Item=(bazel_protos::bytestream::WriteRequest, grpcio::WriteFlags)>>>, [closure@src/store.rs:617:51: 623:14 fingerprint:_]>>, futures::MapErr<grpcio::ClientCStreamReceiver<bazel_protos::bytestream::WriteResponse>, [closure@src/store.rs:625:32: 631:16 fingerprint:_]>, [closure@src/store.rs:624:23: 632:14 receiver:_, fingerprint:_]>, std::result::Result<store::Digest, std::string::String>, [closure@src/store.rs:633:23: 644:14 len:_, fingerprint:_]> : boxfuture::Boxable`

error: aborting due to previous error
```

So then I tried casting instead of ascribing:
```
let stream = futures::stream::iter_ok(bytes)... as Stream<Item=(bazel_protos::bytestream::WriteRequest, grpcio::WriteFlags), Error=grpcio::Error>;
```
which complains that I'm casting to an unsized type, but if I box, it complains the same was as if I try to ascribe as a `Box<Stream<...>>`.

I guess maybe `impl Trait` will make this go away, but in the mean time, it seems that there is a very specific thing I needed to do here, and the tooling's hints and messages guided me in a different direction entirely.

Then again, the typechecker *should* be able to infer the correct type here anyway, because the thing consuming the `Stream` is hard-coded to require a `grpcio::Error`: https://github.com/pingcap/grpc-rs/blob/master/src/call/client.rs#L295

Reply:
> I guess maybe impl Trait will make this go away, but in the mean time, it seems that there is a very specific thing I needed to do here, and the tooling's hints and messages guided me in a different direction entirely.

Yea, that seems correct.

My understanding of why nothing but a type-ascription on the function-call _inputs_ would work, is that that is the only position in which parameters can be generic/abstract (in the absence of `impl Trait` or boxing). With `impl Trait` for local variables, you _maybe_ could have type ascribed this as:
```
let stream: impl Stream<Item=_, Error=grpcio::Error> = ...
```

I've head `impl Trait` referred to as allowing for an "abstract return type", which is not quite the same as having a "generic return type". The caller of a method chooses a generic parameter, but the implementer of a method chooses an abstract return type.

----

As to how to report this one: I think line 624 is maybe not _entirely_ unrelated: the `send_all` call consumes the `stream` variable, and then the `map_err` doesn't add any additional information (because it is satisfied for any error type which implements `Debug`. The `and_then` call doesn't have a bound on the error type, but it does reference it.

Reply:
As to using `future::stream::iter_ok(bytes).chunks(..)`, I'm going to guess that converting into a `Vec<Vec<u8>>` first, and then consuming that with `future::stream::iter_ok` would be more efficient, as it avoids ever having a `Stream<u8, _>`. So: `future::stream::iter_ok(bytes.chunks(..))`. But I haven't profiled, so who knows!

Reply:
I was trying to avoid a `Vec<Vec<u8>>` because I don't want to keep the (potentially large) file being uploaded in memory twice (once for the original `Vec<u8>`, once for the `Vec<Vec<u8>>`), as well as 2 copies of the N chunks currently in flight in both the unserializezd proto and the serialized proto.

If I do `future::stream::iter_ok::<_, grpcio::Error>(bytes.chunks(self.chunk_size_bytes))`, I get complaints that `bytes` doesn't live long enough, because nothing takes ownership of it, so it gets dropped at the end of `store_bytes`. I tried writing a custom `Iterator` implementation which holds on to the underlying `Vec`, but couldn't make it work with as much work as I could justify.

But this is a good catch - I did a little benchmarking, and the difference is massive; chunking before putting into a stream is about 15% faster. I switched to do an eager copy of the chunks, but would love some help somehow passing ownership of the original `Vec` so that the chunking can happen lazily.